src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java

```diff
@@ -198,229 +202,260 @@ public boolean isSigned(String jwt) {
+        Claims claims = parseClaims(payload);
 
-                if (delimiterCount == 0) {
-                    base64UrlEncodedHeader = token;
-                } else if (delimiterCount == 1) {
-                    base64UrlEncodedPayload = token;
-                }
+        // =============== Signature =================
+        validateSignature(jwtParts, header, payload, claims);
 
-                delimiterCount++;
-                sb.setLength(0);
-            } else {
-                sb.append(c);
-            }
-        }
+        //since 0.3:
+        validateClaims(header, claims);
 
-        if (delimiterCount != 2) {
-            String msg = "JWT strings must contain exactly 2 period characters. Found: " + delimiterCount;
-            throw new MalformedJwtException(msg);
-        }
-        if (sb.length() > 0) {
-            base64UrlEncodedDigest = sb.toString();
-        }
+        Object body = claims != null ? claims : payload;
 
-        if (base64UrlEncodedPayload == null) {
-            throw new MalformedJwtException("JWT string '" + jwt + "' is missing a body/payload.");
+        if (jwtParts.getBase64UrlEncodedSignature() != null) {
+            return new DefaultJws<Object>((JwsHeader) header, body, jwtParts.getBase64UrlEncodedSignature());
+        } else {
+            return new DefaultJwt<Object>(header, body);
         }
+    }
 
-        // =============== Header =================
-        Header header = null;
+	private void validateSignature(JwtParts jwtParts, Header header, String payload, Claims claims) {
+		if (!this.ignoreSignature && jwtParts.getBase64UrlEncodedSignature() != null) { //it is signed - validate the signature
 
-        CompressionCodec compressionCodec = null;
+            SignatureAlgorithm algorithm = parseSignatureAlgorithm(header);
 
-        if (base64UrlEncodedHeader != null) {
-            String origValue = TextCodec.BASE64URL.decodeToString(base64UrlEncodedHeader);
-            Map<String, Object> m = readValue(origValue);
+            checkSigningKey();
 
-            if (base64UrlEncodedDigest != null) {
-                header = new DefaultJwsHeader(m);
-            } else {
-                header = new DefaultHeader(m);
-            }
+            //digitally signed, let's assert the signature:
+            Key key = resolveKey(header, payload, claims, algorithm);
 
-            compressionCodec = compressionCodecResolver.resolveCompressionCodec(header);
+            validateSignatureWithSignatureValidator(jwtParts, algorithm, key);
+        }
+	}
+
+	private void validateClaims(Header header, Claims claims) {
+		if (claims != null) {
+
+			if (!this.ignoreExpiry) {
+	            //https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-30#section-4.1.4
+	            //token MUST NOT be accepted on or after any specified exp time:
+	            ValidationHelper.validateExpiration(header, claims, this.allowedClockSkewMillis, this.clock);
```

Comment:
No static/helper methods in the implementation please.  Such methods are not object-oriented - they are more procedural and harder to test.